File: 프로그래머스/2/49994. 방문 길이/방문 길이.cpp
#include <string>
#include <string.h>
#include <map>

#define ROW 6
#define COL 6

using namespace std;

typedef struct {
    int x, y;
} coor;

int solution(string dirs) {
    int answer = 0;
    pair<int, int> s = {0, 0};

    map<pair<pair<int, int>, pair<int, int>>, int> donePath;
    map<char, int> map = {{'U', 0}, {'D', 1}, {'L', 2}, {'R', 3}};
    coor mov[4] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    int donePoint[12][12];
    memset(donePoint, 0, sizeof(donePoint));

    for (const auto& route : dirs) {
        char dir = map[route];
        coor next = {s.first + mov[dir].x, s.second + mov[dir].y};
        if ((next.x < -5 || ROW <= next.x || next.y < -5 || COL <= next.y)) {
            continue;
        }
        if (!donePath[{{s.first, s.second}, {next.x, next.y}}]) {
            answer++;
            donePath[{{s.first, s.second}, {next.x, next.y}}]++;
            donePath[{{next.x, next.y}, {s.first, s.second}}]++;
        }
        s = {next.x, next.y};
    }
    
    return answer;
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.*;

class Solution {
    
    
    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, 1, 0, -1};
    
    
    public int solution(String dirs) {
        int answer = 0;
        
        int nx = 5;
        int ny = 5;
        Set<String> set = new HashSet<>();
        
        for (char c : dirs.toCharArray()) {
            int i = getDir(c);
            
            int bx = nx;
            int by = ny;
            nx += dx[i];
            ny += dy[i];
            
            if (nx < 0 || ny < 0 || nx > 10 || ny > 10) {
                nx = bx;
                ny = by;
                continue;
            }
            set.add("" + bx + by + nx + ny);
            set.add("" + nx + ny + bx + by);
        }
        
        return set.size() / 2;
    }
    
    private static int getDir(char c) {
        if (c == 'U') return 0;
        if (c == 'R') return 1;
        if (c == 'D') return 2;
        return 3;
    }
}